core/src/main/scala/besom/internal/codecs.scala

```diff
@@ -1027,91 +1124,110 @@ object ArgsEncoder:
               .toList
           }
           .map(_.toMap)
-          .map { serializedMap =>
-            val (mapOfResources, mapOfValues) =
-              serializedMap.foldLeft[(Map[String, Set[Resource]], Map[String, Value])](Map.empty -> Map.empty) {
-                case ((mapOfResources, mapOfValues), (label, (resources, value))) =>
-                  if filterOut(label) then (mapOfResources, mapOfValues) // skip filtered
-                  else if Encoder.isEmptySecretValue(value) then (mapOfResources, mapOfValues) // skip nulls in secrets
-                  else if value.kind.isNullValue then
-                    (mapOfResources, mapOfValues) // we treat properties with NullValue values as if they do not exist
-                  else (mapOfResources + (label -> resources), mapOfValues + (label -> value))
+          .flatMap { serializedMap =>
+            val (mapOfMetadata: Map[String, Metadata], mapOfValues: Map[String, Value]) =
+              serializedMap.foldLeft[(Map[String, Metadata], Map[String, Value])](Map.empty -> Map.empty) {
+                case ((mapOfMetadata, mapOfValues), (label, (metadata, value))) =>
+                  if filterOut(label) then (mapOfMetadata, mapOfValues) // skip filtered
+                  else if metadata.empty then
+                    (mapOfMetadata, mapOfValues) // we treat properties with NullValue values as if they do not exist
+                  else (mapOfMetadata + (label -> metadata), mapOfValues + (label -> value))
               }
 
-            val struct = mapOfValues.asStruct
-
-            mapOfResources -> struct
+            if (mapOfMetadata.keySet != mapOfValues.keySet)
+              Result.fail(
+                Exception(
+                  s"Post-condition: expected ArgsEncoder to return metadata and values for the same keys, " +
+                    s"got: ${mapOfMetadata.keys.mkString(",")}, ${mapOfValues.keys.mkString(",")}"
+                )
+              )
+            else
+              Result.pure(mapOfMetadata -> mapOfValues.asStruct)
           }
+end ArgsEncoder
 
-  inline def derived[A <: Product](using m: Mirror.ProductOf[A]): ArgsEncoder[A] =
+// ProviderArgsEncoder and ArgsEncoder are nearly the same with the small difference of
+// ProviderArgsEncoder all fields in being serialized to JSON strings
```

Comment:
same nitpick about not misinforming in comments regarding which fields get JSONed in ProviderArgsEncoder

Reply:
good catch, thanks, done